install/sles-installation-guide.md

```diff
@@ -0,0 +1,47 @@
+# Installing Kata Containers on SLES
+
+Note: Kata Containers is available for SLES version 12 SP3.
+
+This step is only required in case Docker is not installed on the system.
+1. Install the latest version of Docker with the following commands:
+
+```bash
+$ sudo zypper -n install libcgroup1
+$ curl -OkL https://yum.dockerproject.org/repo/main/opensuse/13.2/Packages/docker-engine-17.05.0.ce-1.x86_64.rpm
+$ sudo rpm -ivh docker-engine-17.05.0.ce-1.x86_64.rpm
```

Comment:
- Hard-coded architecture: if this install guide only works for x86_64, this needs to be specified clearly at the top. In fact, if you added the comment and a setup section, you could you do something like:
  ```
      ```bash
      [ $(arch) != "x86_64" ] && echo ERROR: unsupported architecture && exit 1
      ```

  ```

  That would allow https://github.com/kata-containers/tests/blob/master/.ci/kata-doc-to-script.sh to do the right thing (TM) :)

- This looks fragile to me as it's manually installing an OpenSuSE package on SLES. With Clear Containers we simply installed docker - what's the reason we can no longer do that? Is the version too old?
  - https://github.com/clearcontainers/runtime/blob/master/installation/sles-setup.sh#L128

Reply:
@jodh-intel , there is an issue with SLES when we installed docker with
```
$ zypper -n install docker
```
This will install docker version 17.09.1-ce, build f4ffd2511ce9. However, after doing the installation of kata-runtime, kata-proxy, etc which is done correctly and trying to do the following
```
$ sudo systemctl daemon-reload
$ sudo systemctl restart docker
```
We have the following error
```
-- Subject: Unit docker.service has begun start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has begun starting up.
may 22 18:33:44 gaby-sles dockerd[20809]: time="2018-05-22T18:33:44Z" level=info msg="SUSE:secrets :: enabled"
may 22 18:33:44 gaby-sles dockerd[20809]: time="2018-05-22T18:33:44.722479100Z" level=debug msg="Listener created for HTTP on unix (/var/run/docker.sock)"
may 22 18:33:44 gaby-sles dockerd[20809]: Failed to connect to containerd. Please make sure containerd is installed in your PATH or you have specified the correct address. Got error: exec: "docker-containerd": executable file not found in $PATH
may 22 18:33:44 gaby-sles systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
may 22 18:33:44 gaby-sles systemd[1]: Failed to start Docker Application Container Engine.
-- Subject: Unit docker.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit docker.service has failed.
```
Now containerd is being installed by SLES.
containerd version 0.2.3 commit: 06b9cb35161009dcb7123345749fef02f7cea8e0

And docker-containerd does not exist
```
$ zypper search docker*
Loading repository data...
Reading installed packages...
i+ | docker                   
    | docker                      
    | docker-distribution       
    | docker-distribution-registry 
    | docker-image-migrator  
    | docker-img-store-setup 
i   | docker-libnetwork      
i   | docker-runc                 
```
However, this error did not happen when I am trying to manually installed and OpenSuSE package on SLES.

Reply:
Hi @ajaeger - do you have any input on this SLES issue?

Reply:
Hi @GabyCT, @jodh-intel,

I am co-maintaing the Docker packages at SUSE and @ajaeger asked me to have a look at the SLES issue.

> However, this error did not happen when I am trying to manually installed and OpenSuSE package on SLES.

Can you elaborate a bit more which setup is working and which not? In theory, a `zypper install docker` and `systemctl start docker` should always lead to a running Docker daemon on our supported systems (i..e, latest SLES and openSUSE).

> And docker-containerd does not exist

That's correct as we ship it in the `containerd` package, both on SLES and openSUSE.

> may 22 18:33:44 gaby-sles dockerd[20809]: Failed to connect to containerd. Please make sure containerd is installed in your PATH or you have specified the correct address. Got error: exec: "docker-containerd": executable file not found in $PATH

The error log shows that dockerd tries to start docker-containerd and I understand the confusion as we don't ship a `docker-containerd` binary but only a `containerd` one. The reasoning behind is that we manage our services via systemd to avoid containerd being a subprocess of dockerd. We consider this setup to a bit cleaner but it is has no functional impact.

@GabyCT If possible, could you upload the logs of docker and containerd (`journalctl -u docker` and `journalctl -u containerd`)? Can you try to start containerd (`systemctl start containerd`) and check if it's running (`systemctl status containerd`)? Did you change any systemd service files for either docker or containerd?

Reply:
Hi @GabyCT - please can you look at providing these details today if at all possible?

Reply:
Hi @GabyCT - please could you look at this.

/cc @egernst.

Reply:
@jodh-intel sorry, I just saw the comments today. I will take a look and attach all the logs

Reply:
@vrothberg , using the following setup
```
$  cat /etc/os-release
NAME="SLES"
VERSION="12-SP3"
VERSION_ID="12.3"
PRETTY_NAME="SUSE Linux Enterprise Server 12 SP3"
ID="sles"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:12:sp3"
```
The first thing that I did was installed `docker`
```
sudo zypper -n install docker
Refreshing service 'SMT-http_smt-azure_susecloud_net'.
Refreshing service 'cloud_update'.
Loading repository data...
Reading installed packages...
Resolving package dependencies...

The following package is going to be upgraded:
  docker

1 package to upgrade.
Overall download size: 12.3 MiB. Already cached: 0 B. After the operation,
additional 9.0 B will be used.
Continue? [y/n/...? shows all options] (y): y
Retrieving package docker-17.09.1_ce-98.15.1.x86_64
                                           (1/1),  12.3 MiB ( 60.0 MiB unpacked)
Retrieving: docker-17.09.1_ce-98.15.1.x86_64.rpm .........................[done]
Checking for file conflicts: .............................................[done]
(1/1) Installing: docker-17.09.1_ce-98.15.1.x86_64 .......................[done]
Additional rpm output:
Updating /etc/sysconfig/docker...
```

Now checking the status for `docker`
```
sudo systemctl status docker
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; disabled; vendor preset: disabled)
   Active: active (running) since jue 2018-08-02 15:24:14 UTC; 17s ago
     Docs: http://docs.docker.com
 Main PID: 17150 (dockerd)
    Tasks: 11
   Memory: 36.6M
      CPU: 331ms
   CGroup: /system.slice/docker.service
           └─17150 /usr/bin/dockerd --containerd /run/containerd/containerd.sock --add-runtime oci=/usr/sbin/docker-runc
```

Now, I did the installation for `kata-runtime` with the rest of the packages and made the configuration file
```
cat /etc/systemd/system/docker.service.d/kata-containers.conf
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd -D --add-runtime kata-runtime=/usr/bin/kata-runtime --default-runtime=kata-runtime
```
However, when I tried to run `docker` I am getting the failures. Here are the logs

For `docker`
```
$ sudo journalctl -u docker
-- Logs begin at jue 2018-08-02 15:17:55 UTC, end at jue 2018-08-02 15:32:12 UTC. --
ago 02 15:24:13 gabysles systemd[1]: Starting Docker Application Container Engine...
ago 02 15:24:13 gabysles dockerd[17150]: time="2018-08-02T15:24:13Z" level=info msg="SUSE:secrets :: enabled"
ago 02 15:24:13 gabysles dockerd[17150]: time="2018-08-02T15:24:13.482225200Z" level=warning msg="failed to rename /var/lib/docker/tmp for background deletion: rename /var/lib/docker/tmp /var/lib/docker/tmp-old: no such file or directory. Deleting synchronously"
ago 02 15:24:13 gabysles dockerd[17150]: time="2018-08-02T15:24:13.897049500Z" level=info msg="Graph migration to content-addressability took 0.00 seconds"
ago 02 15:24:13 gabysles dockerd[17150]: time="2018-08-02T15:24:13.897451900Z" level=warning msg="Your kernel does not support swap memory limit"
ago 02 15:24:13 gabysles dockerd[17150]: time="2018-08-02T15:24:13.897492300Z" level=warning msg="Your kernel does not support kernel memory limit"
ago 02 15:24:13 gabysles dockerd[17150]: time="2018-08-02T15:24:13.897979600Z" level=info msg="Loading containers: start."
ago 02 15:24:14 gabysles dockerd[17150]: time="2018-08-02T15:24:14.086773300Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
ago 02 15:24:14 gabysles dockerd[17150]: time="2018-08-02T15:24:14.178921300Z" level=info msg="Loading containers: done."
ago 02 15:24:14 gabysles dockerd[17150]: time="2018-08-02T15:24:14.464330800Z" level=info msg="Docker daemon" commit=f4ffd2511ce9 graphdriver(s)=overlay2 version=17.09.1-ce
ago 02 15:24:14 gabysles dockerd[17150]: time="2018-08-02T15:24:14.464497700Z" level=info msg="Daemon has completed initialization"
ago 02 15:24:14 gabysles systemd[1]: Started Docker Application Container Engine.
ago 02 15:24:14 gabysles dockerd[17150]: time="2018-08-02T15:24:14.501539100Z" level=info msg="API listen on /var/run/docker.sock"
ago 02 15:27:54 gabysles systemd[1]: Stopping Docker Application Container Engine...
ago 02 15:27:54 gabysles dockerd[17150]: time="2018-08-02T15:27:54.726003700Z" level=info msg="Processing signal 'terminated'"
ago 02 15:27:54 gabysles systemd[1]: Stopped Docker Application Container Engine.
ago 02 15:27:54 gabysles systemd[1]: Starting Docker Application Container Engine...
ago 02 15:27:54 gabysles dockerd[17598]: time="2018-08-02T15:27:54Z" level=info msg="SUSE:secrets :: enabled"
ago 02 15:27:54 gabysles dockerd[17598]: time="2018-08-02T15:27:54.774235000Z" level=debug msg="Listener created for HTTP on unix (/var/run/docker.sock)"
ago 02 15:27:54 gabysles dockerd[17598]: Failed to connect to containerd. Please make sure containerd is installed in your PATH or you have specified the correct address. Got error: exec: "docker-containerd": executable file not found in $PATH
ago 02 15:27:54 gabysles systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
ago 02 15:27:54 gabysles systemd[1]: Failed to start Docker Application Container Engine.
ago 02 15:27:54 gabysles systemd[1]: docker.service: Unit entered failed state.
ago 02 15:27:54 gabysles systemd[1]: docker.service: Failed with result 'exit-code'.
```

I check the status for `containerd`
```
$ sudo systemctl status containerd
● containerd.service - Containerd Standalone OCI Container Daemon
   Loaded: loaded (/usr/lib/systemd/system/containerd.service; disabled; vendor preset: disabled)
   Active: active (running) since jue 2018-08-02 15:24:13 UTC; 11min ago
     Docs: https://containerd.tools/
 Main PID: 17147 (containerd)
    Tasks: 12
   Memory: 21.2M
      CPU: 272ms
   CGroup: /system.slice/containerd.service
           └─17147 /usr/sbin/containerd --listen fd:// --start-timeout=2m

ago 02 15:24:13 gabysles systemd[1]: Started Containerd Standalone OCI Container Daemon.
```
And I tried to retrieve the logs for `containerd`
```
$ sudo journalctl -u containerd
-- Logs begin at jue 2018-08-02 15:17:55 UTC, end at jue 2018-08-02 15:34:30 UTC. --
ago 02 15:24:13 gabysles systemd[1]: Started Containerd Standalone OCI Container Daemon.
```

Reply:
@jodh-intel ^^^^ here it is the information

Reply:
Thanks @GabyCT.

Reply:
@GabyCT @jodh-intel Thanks for providing the details.

The problem stems from how docker is started:
```
cat /etc/systemd/system/docker.service.d/kata-containers.conf
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd -D --add-runtime kata-runtime=/usr/bin/kata-runtime --default-runtime=kata-runtime
```
On SUSE and openSUSE, we decided to ship containerd as a separate service to decouple docker and containerd a bit more. The containerd binary is thus also only named `containerd` and is placed in `/usr/sbin/containerd`. This explains the error message we see from dockerd: containerd is not running, dockerd tries to execute it but doesn't find the `docker-containerd` binary.

What I suggest to do is to start docker via `systemctl start docker` and rather update `/etc/sysconfig/docker` and add `"-D --add-runtime kata-runtime=/usr/bin/kata-runtime --default-runtime=kata-runtime"` to the `DOCKER_OPTS` variable, which will be added to the set of flags to start docker via `systemctl start docker`.

Would that work for you?

Reply:
@GabyCT I cannot reproduce the issue. Maybe, dockerd was still running when `systemctl start docker` has been invoked?

What I did on my machine is to install the runtime from https://github.com/kata-containers/runtime (`make`, `make install`), then I stopped docker and containerd via systemctl,   added the missing pieces to `/etc/sysconfig/docker` (just with an altered path to kata-runtime). After that I started docker via systemctl and it correctly reports the kata-runtime:
```
Runtimes: kata-runtime oci runc
Default Runtime: kata-runtime
Init Binary: docker-init
```

Can you take another look if the order of those steps works for you? If not, we can set up a call?

Reply:
@vrothberg sorry I am having the same issue, the same failure. Maybe @jodh-intel can try to reproduce the issue in order to see if he is having the same errors.

Reply:
@GabyCT Are you using the docker and containerd packages provided by SUSE? It's just a wild guess, but maybe there were some packages left in your test environment that did not come from SUSE (i.e., `zypper in docker`).

Reply:
@vrothberg, I just did the `zypper -n install docker`

Reply:
@chavafg could you please help me to reproduce this error? I want to see if it is something from my environment or you also hit the same error thanks

Reply:
@GabyCT 
I tried today and I could start the docker daemon with this config:

```
fuentess@sles-kata:~> cat /etc/systemd/system/docker.service.d/kata-containers.conf
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd -D --containerd /run/containerd/containerd.sock --add-runtime kata-runtime=/usr/bin/kata-runtime --default-runtime=kata-runtime
```


I think that @vrothberg approach should also work, but you'll would need to delete the `/etc/systemd/system/docker.service.d/kata-containers.conf` 

I used the docker that is provided by SUSE. Version is: 17.09.1-ce